### PR TITLE
Be able to open zim archive from several fds.

### PR DESCRIPTION
**Pull request by mgautierfr**
Fix https://github.com/openzim/libzim/issues/841

The tests in this PR need a update of zim-testing-suite (https://github.com/openzim/zim-testing-suite/pull/7)

### CODE REVIEW

File: include/zim/archive.h [hunk 1]
```diff
@@ -108,16 +110,29 @@ namespace zim
...
+       *  Construct an archive from several file descriptors.
+       *  Each part may be embedded in a file.
+       *  Fds are used only at Archive creation.
+       *  Ownership of the fds is not taken and they must be closed by caller.
+       *
+       *  Note: This function is not available under Windows.
+       *
+       *  @param fds A vector of FdInput (tuple) containing the fd (int), offset (offset_type) and size (size_type)
+       *             referencing a series of segments representing a complete ZIM archive.
+       */
```
Comment (veloman-yunkan) [inline include/zim/archive.h:134]: I think a separate note will be useful on the situation when multiple segments belong to the same file. As far as I understand, providing access to them via the same shared fd is allowed.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/fileimpl.cpp [hunk 2]
```diff
@@ -173,18 +171,13 @@ class Grouping
   {}
 
   FileImpl::FileImpl(int fd, offset_t offset, zsize_t size)
-    : FileImpl(std::make_shared<FileCompound>(fd), offset, size)
+    : FileImpl(std::make_shared<FileCompound>(std::tuple<int, offset_type, size_type>(fd, offset_type(offset.v), size_type(size.v))))
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/fileimpl.cpp]: Done. (Commit rewritten, no fixup)

File: include/zim/archive.h [hunk 3]
```diff
@@ -115,6 +115,18 @@ namespace zim
...
+       *
+       *  Construct an archive from several file descriptors.
+       *  Each part may be embedded in a file.
+       *
+       *  Note: This function is not available under Windows.
+       *
+       *  @param fds A vector of FdInput (tuple) containing the fd (int), offset (offset_type) and size (size_type)
+       *             referencing a series of segments representing a complete ZIM archive.
+       */
+      explicit Archive(std::vector<FdInput> fds);
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply include/zim/archive.h]: Indeed. I'm not sure it worth it but done for consistency with other constructors.

File: src/file_part.h [hunk 4]
```diff
@@ -22,31 +22,44 @@
...
 #include "zim_types.h"
 #include "fs.h"
 
 namespace zim {
 
+/** A part of file.
+ *
+ * `FilePart` references a part(section) of a physical file.
+ * Most of the time, `FilePart` will reference the whole file (m_offset==0 and m_size==m_fhandle->getSize())
+ * but in some situation, it can reference only a part of the file (in case the content is store in a zip archive.)
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/file_part.h]: Done

File: include/zim/archive.h (hunk 1 above)
Comment (mgautierfr) [reply include/zim/archive.h]: Done

**Comment by mgautierfr** [comment]
Ready for a (final) review. Missing coverage is about error handling code, which is difficult to test as we need to generate io errors.

File: include/zim/zim.h [hunk 5]
```diff
@@ -23,6 +23,7 @@
 #define ZIM_ZIM_H
 
 #include <cstdint>
+#include <tuple>
```
Comment (veloman-yunkan) [inline include/zim/zim.h:26]: Unused header

File: scripts/download_test_data.py [hunk 6]
```diff
@@ -26,7 +26,7 @@
 import tarfile
 import sys
 
-TEST_DATA_VERSION = "0.3"
+TEST_DATA_VERSION = "0.5"
```
Comment (veloman-yunkan) [inline scripts/download_test_data.py:29]: Why don't you merge this commit into the one titled "Add test on multi part embedded file. "?

**Review by veloman-yunkan — APPROVED** [review]
There are a couple of minor suggestions for improvement but nothing bad will happen if the PR is merged as is.

**Comment by mgautierfr** [comment]
Two suggested improvements have been followed. Directly rebased/fixup.
Will merged once CI is ok.

**Pull request merged**